Fix some problems with unexpected behaviour during and after pauses

JonathanSteinbuch opened the pull request:
Hello there,

Pauses have some weird problems. Most notably when you are in the process of building a road when you pause the game then you can click to finish the road and this leaves an unremovable ghost road as seen below.

![pauseProblems](https://user-images.githubusercontent.com/711115/107875860-82142080-6ec2-11eb-8316-2daca846c06a.png)

Another problem is that for many ingame settings windows if you change something during a pause the changes very often are not registered, but still displayed. In the above picture this is visible for the military window, where the number of soldiers in barracks, even after the pause, does not reflect the choice made during the pause.

To address these issues I propose the changes in this pull request.

The guiding principles that I used are as follows (I tried to stay close to the ideas used already in the code and the things decided in some issues):

1. Most importantly at any time the displayed settings should reflect the actual settings or settings that will be changed in the next timer cycle.
2. Since adding game commands in a pause fails (and rightly so, as decided in #1229), any function that issues a game command potentially during a pause has to check whether it was successful and react accordingly.
3. Still a player can click through the windows in pause (to look at stats or see where a road could be built or whatever) and the windows don't have to be aware of the pause. Particularly, I think it's rude to just close the road building window for other players during a pause, so it stays open but they can't finish the road or any other road until the pause ends.
4. If control elements (progress bars and such) can be directly changed during a pause, then if the window is still open when the pause ends, the game commands for these changes will be issued after the pause ends. If the window is closed during a pause however, the noncommited changes made to control elements are discarded.

Especially because of 4 these changes reduce the importance of the visual settings in the game client. So this is related to #350.

During this I removed that for the host the pause starts immediately (as opposed to when the pause command is executed for all players) as it's potentially problematic (for example it's the direct reason for the ghost road problem mentioned at the beginning). If there was a particular reason for the immediate pause for the host we should discuss that.

I hope you find these changes helpful. I think they should make the behaviour during and after pauses much more predictable.

Best Regards
Jonathan

Flamefire reviewed (changes requested):
Great work!
Few notes inline and to your points:
> Most importantly at any time the displayed settings should reflect the actual settings or settings that will be changed in the next timer cycle.

Take care that this might even be later: E.g. changing a setting, that is transmitted and on the way BACK from the server. Then changing it again so the current visual setting will be for the 2nd next NWF tick. From what I've seen this should work ATM.

> Particularly, I think it's rude to just close the road building window for other players during a pause, so it stays open but they can't finish the road or any other road until the pause ends.

See inline comment. Maybe it's better to do so.

> If the window is closed during a pause however, the noncommited changes made to control elements are discarded.

Great that this is defined now. Maybe it makes sense to warn/ask the player about such things? IIRC we have a close function which can prevent closing which could be used to show a message. Sending stuff in the dtor might not have been the best idea anyway...

> During this I removed that for the host the pause starts immediately (as opposed to when the pause command is executed for all players) as it's potentially problematic (for example it's the direct reason for the ghost road problem mentioned at the beginning). If there was a particular reason for the immediate pause for the host we should discuss that.

IMO we should keep and fix that. Especially in a single player game you want immediate pause, it would look strange otherwise. This is very important for replays but maybe also for a single player or during debugging, e.g. pause right before a soldier enters a building. I **think** we could simply call the OnMsgPause function directly instead of setting `framesinfo.isPaused = true;` Could you check that?

JonathanSteinbuch commented on libs/s25main/desktops/dskGameInterface.cpp:
```diff
@@ -930,8 +930,8 @@ void dskGameInterface::Run()
-    // Im Replay und in der Pause keine Straßen bauen
-    if(GAMECLIENT.IsReplayModeOn() || GAMECLIENT.IsPaused())
+    // Im Replay keine Straßen bauen
+    if(GAMECLIENT.IsReplayModeOn())
```
I changed this because I felt it fits better with the behaviour of the other controls. (i.e. you can also look at which buildings to build or at options windows, etc.) And since I added a check that you can only finish building a road outside of a pause this didn't feel necessary anymore.

But if you prefer to continue disallowing this specifically I can reinstate the old version of this line instead

Flamefire replied:
Agreed, ok

JonathanSteinbuch commented on libs/s25main/desktops/dskGameInterface.cpp:
```diff
@@ -1153,18 +1155,6 @@ void dskGameInterface::CI_ReplayEndReached(const std::string& msg)
     messenger.AddMessage(_("SYSTEM"), COLOR_GREY, ChatDestination::System, _("Game was paused."));
-
-    /// Straßenbau ggf. abbrechen, wenn aktiviert
-    if(road.mode != RoadBuildMode::Disabled)
```
You will not be stuck in road mode since you can still manually cancel road building.

That said, my reasoning for removing this is the same as above and if you prefer to have it, no real problem with me.

Flamefire replied:
Ah, ok, then that is fine too.

So only the braces please (in general I avoid braces if the `if` and statement below or short (single line), but here we have a comment and a statement and a very long if so the braces help with readability) -> https://github.com/Return-To-The-Roots/s25client/pull/1372#discussion_r575804883

JonathanSteinbuch commented:
Thank you, @Flamefire, for your helpful review.

> > If the window is closed during a pause however, the noncommited changes made to control elements are discarded.
> 
> Great that this is defined now. Maybe it makes sense to warn/ask the player about such things? IIRC we have a close function which can prevent closing which could be used to show a message. Sending stuff in the dtor might not have been the best idea anyway...

Good idea! I added a dialog box in the case that settings would be discarded. Since it's the same for all of these windows I implemented it in the form of another class TransmittingSettingsWindow with base class IngameWindow.

> > During this I removed that for the host the pause starts immediately (as opposed to when the pause command is executed for all players) as it's potentially problematic (for example it's the direct reason for the ghost road problem mentioned at the beginning). If there was a particular reason for the immediate pause for the host we should discuss that.
> 
> IMO we should keep and fix that. Especially in a single player game you want immediate pause, it would look strange otherwise. This is very important for replays but maybe also for a single player or during debugging, e.g. pause right before a soldier enters a building. I **think** we could simply call the OnMsgPause function directly instead of setting `framesinfo.isPaused = true;` Could you check that?

Yes, this should work. It feels a bit wrong to call OnMsgPause directly, but I did it anyway.

Flamefire commented on libs/s25main/ingameWindows/TransmittingSettingsWindow.cpp:
```diff
@@ -0,0 +1,70 @@
+{
+    // Timer for transmitting changes every 2 seconds
+    using namespace std::chrono_literals;
+    AddTimer(1001, 2s);
```
Not sure here. Some windows assign special values to the IDs e.g. to map goods to ids and back. It should at least be a protected static constexpr value. Best would likely be to give this an ID of 1 or 0(?) and use the enum-id approach with IDs starting after that. But I'd guess that is overkill for now, so at least make it a named constant...

Flamefire commented on libs/s25main/ingameWindows/TransmittingSettingsWindow.h:
```diff
@@ -0,0 +1,44 @@
+                               glArchivItem_Bitmap* background, bool modal = false, bool closeOnRightClick = true,
+                               Window* parent = nullptr);
+
+    virtual ~TransmittingSettingsWindow() {}
```
```suggestion
```

No need, IngameWindow and Window is already virtual

Flamefire commented on libs/s25main/ingameWindows/TransmittingSettingsWindow.h:
```diff
@@ -0,0 +1,44 @@
+    virtual ~TransmittingSettingsWindow() {}
+
+    /// Updates the control elements with values from visual settings
+    virtual void UpdateSettings() {}
+    /// sends potential changes to the client
+    virtual void TransmitSettings() {}
```
```suggestion
    virtual void UpdateSettings() = 0;
    /// sends potential changes to the client
    virtual void TransmitSettings() = 0;
```

I think it doesn't make sense to subclass this without implementing those 2

Flamefire commented on libs/s25main/ingameWindows/TransmittingSettingsWindow.h:
```diff
@@ -0,0 +1,44 @@
+
+    void Close() override;
+
+    virtual void Msg_Timer(unsigned ctrl_id) override;
```
```suggestion
    void Msg_Timer(unsigned ctrl_id) override;
```

Per convention: `virtual` introduces a method and `override [final]` overrides it. So always use either or for consistence

Flamefire commented on libs/s25main/ingameWindows/TransmittingSettingsWindow.h:
```diff
@@ -0,0 +1,44 @@
+    void Close() override;
+
+    virtual void Msg_Timer(unsigned ctrl_id) override;
+    void Msg_MsgBoxResult(const unsigned msgbox_id, const MsgboxResult mbr) override;
```
```suggestion
    void Msg_MsgBoxResult(unsigned msgbox_id, MsgboxResult mbr) override;
```

`const` only in impl. clang-tidy will complain anyway

Flamefire commented on libs/s25main/ingameWindows/iwAction.cpp:
```diff
@@ -338,12 +338,14 @@ void iwAction::AddUpgradeRoad(ctrlGroup* group, unsigned& /*x*/, unsigned& width
-        GAMECLIENT.UpgradeRoad(flag->GetPos(), flag_dir);
+        return GAMECLIENT.UpgradeRoad(flag->GetPos(), flag_dir);
+    else
+        return false;
```
**Maybe**:

```suggestion
        return true;
```

If there is no flag anymore, the task is done and the window can be closed, or what do you think?

JonathanSteinbuch replied:
I think you are exactly right.

Flamefire reviewed (changes requested):
I like the approach, I think it makes sense. If you want a better name, IIRC classes like that were called "Adapter", so maybe `TransmitSettingsIgwAdapter`? (I think we use Igw in other places) Just so it is clear that this is not a real Window yet. But up to you

JonathanSteinbuch commented on libs/s25main/ingameWindows/TransmittingSettingsWindow.cpp:
```diff
@@ -0,0 +1,70 @@
+
+void TransmittingSettingsWindow::Msg_MsgBoxResult(const unsigned msgbox_id, const MsgboxResult mbr)
+{
+    switch(msgbox_id)
```
True :D

JonathanSteinbuch commented on libs/s25main/network/GameClient.cpp:
```diff
@@ -1660,9 +1660,10 @@ void GameClient::SetPause(bool pause)
     } else if(IsHost())
     {
         // Pause instantly
+        auto msg = new GameMessage_Pause(pause);
```
```suggestion
        auto *msg = new GameMessage_Pause(pause);
```

JonathanSteinbuch commented on libs/s25main/network/GameClient.cpp:
```diff
@@ -1660,9 +1660,10 @@ void GameClient::SetPause(bool pause)
     } else if(IsHost())
     {
         // Pause instantly
+        auto *msg = new GameMessage_Pause(pause);
```
```suggestion
        auto* msg = new GameMessage_Pause(pause);
```
